harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -28,56 +31,199 @@
     pass
 
 
-class PyTorchTrialController(det.TrialController):
-    def __init__(self, trial_inst: det.Trial, *args: Any, **kwargs: Any) -> None:
-        super().__init__(*args, **kwargs)
+class TrainUnit:
+    def __init__(self, value: int):
+        self.value = value
+
+    @staticmethod
+    def _from_searcher_unit(length: int, unit: core.Unit):
+        if unit == core.Unit.EPOCHS:
+            return Epoch(length)
+        elif unit == core.Unit.RECORDS:
+            return Record(length)
+        elif unit == core.Unit.BATCHES:
+            return Batch(length)
+        else:
+            raise ValueError(f"unrecognized searcher unit {unit}")
+
+    @staticmethod
+    def _from_values(
+        batches: Optional[int] = None, records: Optional[int] = None, epochs: Optional[int] = None
+    ):
+        if sum((batches is not None, records is not None, epochs is not None)) != 1:
+            raise ValueError(f"invalid length: batches={batches} records={records} epochs={epochs}")
+        if batches:
+            return Batch(batches)
+        elif records:
+            return Record(records)
+        elif epochs:
+            return Epoch(epochs)
+
+    def _divides(self, steps: int):
+        assert self.value > 0, "TrainUnit value must be > 0"
+        if steps == 0:
+            return False
+        return steps % self.value == 0
+
+
+class Epoch(TrainUnit):
+    pass
+
+
+class Batch(TrainUnit):
+    pass
+
+
+class Record(TrainUnit):
+    pass
+
+
+class _TrainStepType(enum.Enum):
+    CHECKPOINT = "CHECKPOINT"
+    REPORT = "REPORT"
+    VALIDATE = "VALIDATE"
+    TRAIN = "TRAIN"
+
+
+# XXX: should this be private? will be returned from .train() but is only used internally
```

Comment:
yes it should be private, at least until we decide that `.train()` should be public